stream_alert/alert_processor/outputs.py

```diff
@@ -556,6 +564,88 @@ def dispatch(self, **kwargs):
 
 
 @output
+class KinesisFirehoseOutput(AWSOutput):
+    """High throughput Alert delivery to AWS S3"""
+    MAX_RECORD_SIZE = 1000 * 1000
+    MAX_BACKOFF_ATTEMPTS = 3
+
+    __service__ = 'aws-firehose'
+    __aws_client__ = None
+
+    def get_user_defined_properties(self):
+        """Properties asssigned by the user when configuring a new Firehose output
+
+        Every output should return a dict that contains a 'descriptor' with a description of the
+        integration being configured.
+
+        Returns:
+            OrderedDict: Contains various OutputProperty items
+        """
+        return OrderedDict([
+            ('descriptor',
+             OutputProperty(
+                 description='a short and unique descriptor for this Firehose Delivery Stream')),
+            ('aws_value',
+             OutputProperty(description='the Firehose Delivery Stream name'))
+        ])
+
+    def dispatch(self, **kwargs):
+        """Send alert to a Kinesis Firehose Delivery Stream
+
+        Keyword Args:
+            descriptor (str): Service descriptor (ie: slack channel, pd integration)
+            rule_name (str): Name of the triggered rule
+            alert (dict): Alert relevant to the triggered rule
+
+        Returns:
+            bool: Indicates a successful or failed dispatch of the alert
+        """
+        @backoff.on_exception(backoff.fibo,
+                              ClientError,
+                              max_tries=self.MAX_BACKOFF_ATTEMPTS,
+                              jitter=backoff.full_jitter,
+                              on_backoff=backoff_handler,
+                              on_success=success_handler,
+                              on_giveup=giveup_handler)
+        def _firehose_request_wrapper(json_alert, delivery_stream):
+            """Make the PutRecord request to Kinesis Firehose with backoff
+
+            Args:
+                json_alert (str): The JSON dumped alert body
+                delivery_stream (str): The Firehose Delivery Stream to send to
+
+            Returns:
+                dict: Firehose response in the format below
+                    {'RecordId': 'string'}
+            """
+            return self.__aws_client__.put_record(DeliveryStreamName=delivery_stream,
```

Comment:
Should this be in try/except that catches `ClientError` and returns something falsey?

Reply:
nope - backoff handles it

Reply:
by using `backoff.on_exception` you want the wrapped function to raise, or else backoff won't happen

Reply:
oh right I forgot that. so what will this return if there there are 3 ClientErrors in a row (ie max attempts are reached)? none? the 4th clienterror? 🤔

Reply:
it raises, shouldn't [the handler](https://github.com/airbnb/streamalert/blob/master/stream_alert/alert_processor/main.py#L112-L120) catch it?

Reply:
yeah it will - just wanted to make sure that was expected, sgtm!